proc/cut.go

```diff
@@ -5,74 +5,268 @@ import (
+			// is currently working on, record is the nested
+			// record for the current field.  First figure out
+			// what (if any) common parents are shared.
+			l := len(currentRecord)
+			if len(record) < l {
+				l = len(record)
+			}
+			pos := 0
+			for pos < l {
+				if record[pos] != currentRecord[pos] {
+					break
+				}
+				pos += 1
+			}
+
+			// Note any previously encoded records that are
+			// now finished.
+			if i > 0 {
+				fieldInfos[i-1].containerEnds = len(currentRecord) - pos
+			}
+
+			// Validate any new records that we're starting
+			// (i.e., ensure that we didn't handle fields from
+			// the same record previously), then record the names
+			// of all these records.
+			for pos2 := pos; pos2 < len(record); pos2++ {
+				recname := strings.Join(record[:pos2+1], ".")
+				_, seen := seenRecords[recname]
+				if seen {
+					return nil, fmt.Errorf("All cut fields in record %s must be adjacent", recname)
+				}
+				seenRecords[recname] = true
+				containerBegins = append(containerBegins, record[pos2])
+			}
+			currentRecord = record
+		}
+		fullname := strings.Join(names, ".")
+		fname := names[len(names)-1]
+		fieldInfos = append(fieldInfos, fieldInfo{resolver, fname, fullname, containerBegins, 0})
+	}
+	fieldInfos[len(fieldInfos)-1].containerEnds = len(currentRecord)
+
 	return &Cut{
-		Base:   Base{Context: c, Parent: parent},
-		fields: fields,
-		cutmap: make(map[int]*zson.Descriptor),
+		Base:    Base{Context: c, Parent: parent},
+		fields:  fieldInfos,
+		cutmap:  make(map[int]*zson.Descriptor),
+		builder: zval.NewBuilder(),
+	}, nil
+}
+
+// Split an ast.FieldExpr representing a chain of record field references
+// into a list of strings representing the names.
+// E.g., "x.y.z" -> ["x", "y", "z"]
+func split(node ast.FieldExpr) ([]string, error) {
+	switch n := node.(type) {
+	case *ast.FieldRead:
+		return []string{n.Field}, nil
+	case *ast.FieldCall:
+		if n.Fn != "RecordFieldRead" {
+			return nil, fmt.Errorf("unexpected field op %s", n.Fn)
+		}
+		names, err := split(n.Field)
+		if err != nil {
+			return nil, err
+		}
+		return append(names, n.Param), nil
+	default:
+		return nil, fmt.Errorf("unexpected node type %T", node)
 	}
 }
 
-func (c *Cut) lookup(in *zson.Descriptor) *zson.Descriptor {
-	d, ok := c.cutmap[in.ID]
-	if ok {
-		return d
+func sameRecord(names1, names2 []string) bool {
+	if len(names1) != len(names2) {
+		return false
 	}
-	var columns []zeek.Column
-	for _, field := range c.fields {
-		colno, ok := in.ColumnOfField(field)
-		if !ok {
-			// a field is missing... block this descriptor
-			c.cutmap[in.ID] = nil
-			c.nblocked++
-			return nil
+	for i := range names1 {
+		if names1[i] != names2[i] {
+			return false
 		}
-		columns = append(columns, in.Type.Columns[colno])
 	}
-	out := c.Resolver.GetByColumns(columns)
-	c.cutmap[in.ID] = out
-	return out
+	return true
 }
 
 // CreateCut returns a new record value derived by keeping only the fields
```

Comment:
since you're in here, s/CreateCut/cut/

Reply:
Will do.  I also notice that this function never returns a non-nil error so I'm going to remove it from the signature.